src/services/photos/sync/photosRemoteSyncManager/photosRemoteSyncManager.ts

```diff
@@ -0,0 +1,201 @@
+  private logger: BaseLogger;
+  private totalPhotosSyncedFromRemote = 0;
+  private lastPhotoPulledDate: Date | null = null;
+  private onStatusChangeCallback: OnStatusChangeCallback = () => {
+    /** NOOP */
+  };
+
+  private onRemotePhotosPageSyncedCallback: OnRemotePhotosPageSyncedCallback = () => {
+    /** NOOP */
+  };
+  public status: PhotosRemoteSyncManagerStatus = PhotosRemoteSyncManagerStatus.IDLE;
+  constructor(private devicePhotosScanner: DevicePhotosScannerService) {
+    this.logger = new BaseLogger({
+      enabled: ENABLE_PHOTOS_REMOTE_SYNC_MANAGER_LOGS,
+      tag: 'PHOTOS_REMOTE_SYNC',
+    });
+  }
+  private previewsQueue = async.queue<{ photo: Photo; retries: number }, Photo, Error>(async (task, next) => {
+    const startTaskTime = Date.now();
+    const photosItem = photosUtils.getPhotosItem(task.photo);
+
+    // Check if we have the preview locally,
+    // if not download and save it
+    try {
+      await photosRealmDB.savePhotosItem(task.photo, false);
+
+      const existsInDevice = this.devicePhotosScanner.getPhotoInDevice(photosItem.name, photosItem.takenAt)
+        ? true
+        : false;
+
+      if (existsInDevice) {
+        this.logger.info(
+          `Photo exists in device, skipping preview download, task done in ${Date.now() - startTaskTime}ms`,
+        );
+
+        // We are done, save the date
+        await asyncStorageService.saveLastPhotoPulledDate(task.photo.updatedAt);
+        next(null, task.photo);
+
+        // Exists in device, add the photo
+        this.totalPhotosSyncedFromRemote++;
+        return;
+      }
+    } catch (err) {
+      this.logger.error(`Error preparing preview ${(err as Error).message}`);
+      next(err as Error);
+      return;
+    }
+
+    const existsPreviewFile = await fs.exists(photosItem.localPreviewPath);
+
+    try {
+      if (!existsPreviewFile) {
+        this.logger.info('Missing preview, downloading it');
+        await photosPreview.getPreview(photosItem);
+      }
+      // Downloaded already, add photo
+      this.totalPhotosSyncedFromRemote++;
+
+      this.logger.info(`Preview downloaded, task done in ${Date.now() - startTaskTime}ms`);
+      // We are done, save the date
+      await asyncStorageService.saveLastPhotoPulledDate(task.photo.updatedAt);
+      return next(null, task.photo);
+    } catch (err) {
+      this.logger.error(`Preview download failed ${(err as Error).message}`);
+
+      // Retry the preview download, or fail, we cannot wait for all the previews
+      // to be downloaded, since we don't save the failed ones, we just retry the task
+      // if no success, we will go to the next Photos page ignoring the failed ones
+      // Note that No preview is not the same as no photo in the view, the photo and the
+      // metadata will appear, but not the preview image. GalleryItem component has a mechanism
+      // to detect failing previews and retry the download
+      if (task.retries < MAX_PREVIEW_DOWNLOAD_RETRIES) {
```

Comment:
Should we retry and stop the next previews to load or just re-queue this task again? A nice discussion can be made here. This failure can make the sync always slower if the same preview needs always all the retries.

Reply:
retrys always go to the end of the queue, so they will be retried once we finish with the other ones